File: stock_search/data_sources/stockanalysis/page_scrapers/statistics.py
"""StockAnalysis statistics page scraping helpers."""

from __future__ import annotations

from collections.abc import Callable
import re

from selectolax.lexbor import LexborHTMLParser

from ..parsing import (
    build_model_from_rows,
    extract_quote_scalar,
    extract_table_rows,
    parse_number,
)
from ..schemas import StockAnalysisStatistics

STOCKANALYSIS_STATISTICS_URL = "https://stockanalysis.com/stocks/{ticker}/statistics/"

QUOTE_BLOCK_PATTERN = re.compile(r"quote:\{(.*?)\},stream:", re.DOTALL)
QUOTE_EMPTY_FIELDS = {"price": None, "change": None, "change_percent_1d": None}
REGULAR_QUOTE_KEYS = {"price": "p", "change": "c", "change_percent_1d": "cp"}
EXTENDED_QUOTE_KEYS = {"price": "ep", "change": "ec", "change_percent_1d": "ecp"}
EXTENDED_SESSION_NAMES = {"Pre-market", "After-hours"}

STATISTICS_FIELD_SPECS = {
    "market_cap": ("Market Cap", "parse_number"),
    "beta": ("Beta (5Y)", "parse_number"),
    "fifty_two_week_price_change": ("52-Week Price Change", "parse_percent_points"),
    "moving_average_50d": ("50-Day Moving Average", "parse_number"),
    "moving_average_200d": ("200-Day Moving Average", "parse_number"),
    "rsi": ("Relative Strength Index (RSI)", "parse_number"),
    "average_volume_20d": ("Average Volume (20 Days)", "parse_number"),
    "pe": ("PE Ratio", "parse_number"),
    "pe_forward": ("Forward PE", "parse_number"),
    "peg": ("PEG Ratio", "parse_number"),
    "roic": ("Return on Invested Capital (ROIC)", "parse_percent_ratio"),
    "gross_margin": ("Gross Margin", "parse_percent_ratio"),
    "operating_margin": ("Operating Margin", "parse_percent_ratio"),
    "debt_to_equity": ("Debt / Equity", "parse_number"),
    "debt_to_ebitda": ("Debt / EBITDA", "parse_number"),
    "free_cash_flow": ("Free Cash Flow", "parse_number"),
}


def _extract_quote_block(
    *,
    ticker_lower: str,
    fetch_html: Callable[[str], str | None],
) -> str | None:
    """Extract the embedded quote block from the statistics page."""
    statistics_url = STOCKANALYSIS_STATISTICS_URL.format(ticker=ticker_lower)
    html = fetch_html(statistics_url)
    if html is None:
        return None

    quote_match = QUOTE_BLOCK_PATTERN.search(html)
    if not quote_match:
        return None
    return quote_match.group(1)


def _extract_quote_values(
    quote_block: str,
    field_keys: dict[str, str],
) -> dict[str, float | None]:
    """Parse a set of quote fields from an embedded quote block."""
    return {field_name: parse_number(extract_quote_scalar(quote_block, key) or "") for field_name, key in field_keys.items()}


def scrape_statistics_snapshot(
    *,
    ticker_lower: str,
    fetch_soup: Callable[[str], LexborHTMLParser | None],
) -> StockAnalysisStatistics:
    """Scrape the StockAnalysis statistics table into a snapshot."""
    statistics_url = STOCKANALYSIS_STATISTICS_URL.format(ticker=ticker_lower)
    soup = fetch_soup(statistics_url)
    if soup is None:
        return StockAnalysisStatistics()

    rows = extract_table_rows(soup, cell_selector="th, td", keep_first=True)
    return build_model_from_rows(
        rows,
        model_type=StockAnalysisStatistics,
        field_specs=STATISTICS_FIELD_SPECS,
    )


def scrape_quote_fields(
    *,
    ticker_lower: str,
    fetch_html: Callable[[str], str | None],
) -> dict[str, float | None]:
    """Extract quote fields embedded on the statistics page."""
    quote_block = _extract_quote_block(ticker_lower=ticker_lower, fetch_html=fetch_html)
    if quote_block is None:
        return dict(QUOTE_EMPTY_FIELDS)

    regular_quote = _extract_quote_values(quote_block, REGULAR_QUOTE_KEYS)
    extended_quote = _extract_quote_values(quote_block, EXTENDED_QUOTE_KEYS)
    extended_session = extract_quote_scalar(quote_block, "es")

    if extended_session in EXTENDED_SESSION_NAMES and extended_quote["price"] is not None:
        return extended_quote
    return regular_quote

File: src/stock-search/data-sources/stockanalysis/page-scrapers/statistics.ts
/** StockAnalysis statistics page scraping helpers. */

import {
	entryValue,
	flattenStatisticsEntries,
	type StockAnalysisStatisticsPayload,
} from "../parsing.js";
import { STOCKANALYSIS_STATISTICS_URL } from "../urls.js";
import {
	evaluateObjectLiteral,
	extractObjectLiteral,
	fetchText,
	parseNumberText,
} from "../../shared.js";

const QUOTE_BLOCK_PATTERN = /quote:\{(.*?)\},stream:/s;
const QUOTE_EMPTY_FIELDS = {
	price: null,
	change: null,
	change_percent_1d: null,
};
const REGULAR_QUOTE_KEYS = {
	price: "p",
	change: "c",
	change_percent_1d: "cp",
} as const;
const EXTENDED_QUOTE_KEYS = {
	price: "ep",
	change: "ec",
	change_percent_1d: "ecp",
} as const;
const EXTENDED_SESSION_NAMES = new Set(["Pre-market", "After-hours"]);

function extractQuoteScalar(quoteBlock: string, key: string): string | null {
	const match = quoteBlock.match(new RegExp(`${key}:("([^"]*)"|[^,}]+)`));
	if (!match) {
		return null;
	}
	const rawValue = match[2] ?? match[1] ?? "";
	return String(rawValue).replace(/^"|"$/g, "").trim() || null;
}

function extractQuoteValues(
	quoteBlock: string,
	fieldKeys: Record<keyof typeof QUOTE_EMPTY_FIELDS, string>,
): Record<keyof typeof QUOTE_EMPTY_FIELDS, number | null> {
	return {
		price: parseNumberText(extractQuoteScalar(quoteBlock, fieldKeys.price)),
		change: parseNumberText(extractQuoteScalar(quoteBlock, fieldKeys.change)),
		change_percent_1d: parseNumberText(
			extractQuoteScalar(quoteBlock, fieldKeys.change_percent_1d),
		),
	};
}

/** Scrape the StockAnalysis statistics page into a structured snapshot. */
export async function scrapeStatisticsSnapshot(
	tickerLower: string,
): Promise<Record<string, unknown>> {
	const html = await fetchText(
		STOCKANALYSIS_STATISTICS_URL.replace("{ticker}", tickerLower),
	);
	const payload = evaluateObjectLiteral<StockAnalysisStatisticsPayload>(
		extractObjectLiteral(html ?? "", "statistics"),
	);
	const entries = flattenStatisticsEntries(payload);
	if (entries.size === 0) {
		return {};
	}

	const medianUpside = entryValue(entries, "priceTargetChange");
	const priceTarget = entryValue(entries, "priceTarget");
	const analystConsensus = entries.get("analystRatings")?.value;
	const analystCount = entryValue(entries, "analystCount");

	return {
		market_cap: entryValue(entries, "marketCap"),
		pe: entryValue(entries, "pe"),
		pe_forward: entryValue(entries, "peForward"),
		peg: entryValue(entries, "pegRatio"),
		beta: entryValue(entries, "beta"),
		free_cash_flow: entryValue(entries, "fcf"),
		gross_margin: entryValue(entries, "grossMargin"),
		debt_to_equity: entryValue(entries, "debtEquity"),
		rsi: entryValue(entries, "rsi"),
		median_upside: medianUpside,
		ratings:
			medianUpside != null || priceTarget != null || analystConsensus != null
				? [
						{
							firm: "Consensus",
							to_grade: analystConsensus ?? null,
							from_grade: null,
							action: "consensus",
							date: null,
							analyst_count: analystCount,
							price_target: priceTarget,
							upside_pct: medianUpside,
						},
					]
		: null,
	};
}

/** Extract quote fields embedded on the statistics page. */
export async function scrapeQuoteFields(
	tickerLower: string,
): Promise<Record<keyof typeof QUOTE_EMPTY_FIELDS, number | null>> {
	const html = await fetchText(
		STOCKANALYSIS_STATISTICS_URL.replace("{ticker}", tickerLower),
	);
	const quoteBlock = html?.match(QUOTE_BLOCK_PATTERN)?.[1];
	if (!quoteBlock) {
		return { ...QUOTE_EMPTY_FIELDS };
	}

	const regularQuote = extractQuoteValues(quoteBlock, REGULAR_QUOTE_KEYS);
	const extendedQuote = extractQuoteValues(quoteBlock, EXTENDED_QUOTE_KEYS);
	const extendedSession = extractQuoteScalar(quoteBlock, "es");

	if (
		extendedSession &&
		EXTENDED_SESSION_NAMES.has(extendedSession) &&
		extendedQuote.price != null
	) {
		return extendedQuote;
	}
	return regularQuote;
}
